layout2020 (flexbox): Implement `start`, `end`, and `space-evenly` content alignment

nicoburns opened the pull request:
<!-- Please describe your changes on the following line: -->
Implements support for `start`, `end` and `space-evenly` values of the `align-` and `justify-` content properties. Also fixes alignment for some of the other values (`space-between`) in some cases. Note that the actual alignment logic is duplicated for `AlignContent` and `JustifyContent`. This is required as they are different types. If we switch to Gecko's alignment implementation then this could be de-duplicated as they become wrappers of a shared underling type.

Depends on https://github.com/servo/stylo/pull/21

### Questions

- [x] Is there an automated way to update expected WPT test results? I have updated them manually for this PR.

### Notes on WPT test changes:

**align-content-wrap-003.html**
- PASS due to added space-evenly support
- PASS due to fixed space-between behaviour when content overflows
- FAIL is genuine according to test. But our result matches both Chrome and Firefox.

**align-content-wrap-005.html**
- FAIL due to missing `direction` support (was previously passing by coincidence)
- FAIL due to missing `direction` support (was previously passing by coincidence)

**flexbox_justifycontent-rtl-001.html**
- PASS due to added space-evenly support
- FAIL due to missing `direction` support (was previously passing by coincidence)

**flexbox-flex-direction-column-percentage-ignored.html**
- This file tests flexbox columns which are entirely unimplemented, so it is expected that this test would fail (and surprising that it was passing before).

**overflow-auto-008.html**
- Suspected spurious error. Off by one and these changes should not affect that test.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are (WPT) tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

nicoburns committed:
Implement start, end, space-evenly content alignment + fix others

Update test expectations for content alignment fixes

Revert test expectations that are still generating the old results in CI

Update layout2013 test expectation for content alignment

Update content alignment fallback to use safe alignment

Implement fallback alignment

Update content alignment with recent spec changes

jdm commented:
https://github.com/servo/servo/blob/main/tests/wpt/README.md#updating-test-expectations should answer your question about updating test expectations.

nicoburns commented:
@jdm Thanks! I've noticed that this PR is failing the "Tidy" CI check due to duplicate entries in the Cargo.lock. Is there a process for updating Cargo.lock to avoid this? (I've updated Cargo.toml to point to temporarily point to my branch of Stylo that this PR depends on - and I deleted and recreated Cargo.toml to make to work but it seems like CI is not even running most of it's checks in the current state).

jdm commented:
You can update https://github.com/servo/servo/blob/main/servo-tidy.toml to allow duplicates.

jdm commented:
It looks like you ran a general cargo update, though, which it would be good to avoid unless it's actually required for these changes.

nicoburns commented:
OK, I've fixed the cargo.lock update to be just Stylo. The WPT check still seems to be skipped (does this need to be triggered manually?), but we seem to be passing the checks that have been run :)

mrobinson commented:
WPT tests are triggered via adding labels. They take too long to run for every push. I've triggered them here.

nicoburns commented:
@mrobinson 

1. Can I get another WPT run on this? I've updated the test expectations match the previous run. These changes were al reverting expectations back to the way they are currently set on `main` so I am expecting that to be uncontroversial.

3. Assuming that there are no "Stable unexpected results" in the next run, what would it take to get this merged?

   I assume that we would need to:
   
     - merge https://github.com/servo/stylo/pull/21 first
     - undo the current changes to `Cargo.toml`/`Cargo.lock` that are in this PR
     - apply some different changes. I guess we would need to land the stylo changes in the `2023-09-01` branch as well as `main` as Servo is currently pointed to that branch, and then update `Cargo.lock` in this PR to point to the new commit.
   
   It seems to me that this process would be easier if we pinned a specific git commit of stylo in servo's `Cargo.toml` (rather than relying on `Cargo.lock` for this - which seems easy to accidentally unpin), but perhaps there is some reason why this can't be done?
   
P.S. What are the criteria for being added to the Servo github organisation / who's decision would that be? It would be good to be able to add the CI-triggering labels myself so I don't have to take up anybody else's time

mrobinson commented:
>     1. Can I get another WPT run on this? I've updated the test expectations match the previous run. These changes were al reverting expectations back to the way they are currently set on `main` so I am expecting that to be uncontroversial.

Apologies for the delay. I've triggered a new WPT run.

>     2. Assuming that there are no "Stable unexpected results" in the next run, what would it take to get this merged?
>        I assume that we would need to:

Someone who understands flex layout or is willing to dive in here, will just need to review these changes. I think we also need to land https://github.com/servo/stylo/pull/21 first.

> P.S. What are the criteria for being added to the Servo github organisation / who's decision would that be? It would be good to be able to add the CI-triggering labels myself so I don't have to take up anybody else's time

It's all a bit informal now, but we should likely add a bit more more process here.

nicoburns commented:
> Stable unexpected results (20)

I think these failures were due to https://github.com/servo/stylo/pull/21 becoming out of date due to https://github.com/servo/stylo/pull/22 being recently landed. I've rebased. I suppose we'll need to do another WPT run?

> Someone who understands flex layout or is willing to dive in here, will just need to review these changes. I think we also need to land https://github.com/servo/stylo/pull/21 first.

@Loirooriol would you be an appropriate person to review this and https://github.com/servo/stylo/pull/21 (which is tiny btw). Also, is there anyone else active in the Servo project who is familiar enough with Flexbox to review such PRs, because I have a bunch more changes like this planned (streaming in flexbox-local changes from well-tested code in Taffy, so I'm pretty confident in the changes, but I imagine Servo will still want them reviewed).

In this case servo `main` actually builds fine with the stylo PR applied (as Servo code has a fallback path in the code that matches on these enums) so it could probably be landed separately. In general there seems to be quite a tricky dance where:

1. The Servo change needs to approved
2. The Style change needs to be approved
3. The Stylo change needs to be landed
4. The Servo PR needs updated to point back to Stylo `main` or dated (e.g. `2023-09-01` branch)
5. The Servo PR then needs to be merged

I suppose once 1. and 2. are completed, the PR author might be able to complete steps 3-5?

nicoburns commented:
Ah, the layout2013 tests. I've pushed an update that updates those expectations (they were all newly passing tests that correspond to the changes in this PR).

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -761,30 +815,68 @@ impl FlexLine<'_> {
+        // (although this will change once gap is implemented)
+        // TODO: Implement gap property
+        let item_main_interval = if free_space_distributed ||
+            item_count == 0 ||
```
Maybe `item_count <= 1`?

nicoburns commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -301,19 +309,65 @@ impl FlexContainer {
+                // Note: free_space may be negative in which case we wish the gaps between items to be zero
+                // (although this will change once gap is implemented)
+                // TODO: Implement gap property
+                let item_main_interval = if free_space <= Length::zero() || line_count == 0 {
```
Yeah, I think we'd better keep the condition here as the line_count >= 2 seems liable to be removed. Happy to change it to `<= 1` if you think that's clearer. Devide by zero is definitely wrong here as we'd end up with `NaN` rather than `0`.

nicoburns commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -301,19 +309,65 @@ impl FlexContainer {
         let line_interval = match flex_context.container_definite_inner_size.cross {
             Some(cross_size) if line_count >= 2 => {
```
Yeah, frankly all sorts of things are wrong with Servo's Flexbox layout. Implementing `align-content: stretch` is on my todo list (would probably make sense to fix this at the same time) behind `flex-direction: column` which is IMO the most egregious missing feature.

nicoburns commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -301,19 +309,65 @@ impl FlexContainer {
+                        if free_space >= Length::zero() {
+                            (free_space / line_count as CSSFloat) / 2.0
+                        } else {
+                            free_space / 2.0
```
Ah, I was aware of the safe alignment `keyword`, but last I had checked the default change hadn't been widely implemented (or checked for web compatibility). Am I right in thinking that "falling back to safe centering" is equivalent to just falling back to `start` (as if we're falling back at all, then `safe center` will itself fall back to `start`)?

If we're going to change the default to safe alignment then I'm thinking we should probably also implement the keywords themselves. Which will likely mean that we want to switch over to Gecko's representation of the alignment styles in Stylo (and we may also want to switch the `-items` and `-self` properties).

My preference would be to land this first (it being an improvement on what we had) and make the larger change as a follow up PR. But if you'd prefer to jump straight to that implementation in one big leap then I could also do that. What do you think?

Loirooriol replied:
> Am I right in thinking that "falling back to safe centering" is equivalent to just falling back to start

I think that's the case. Well, maybe if the flex container is a scroll container, the start scroll position might be different (https://drafts.csswg.org/css-align/#overflow-scroll-position) but I think nobody implements this well.

> But if you'd prefer to jump straight to that implementation in one big leap then I could also do that

No, I didn't mean to make the larger changes here, I just meant using `Length::zero()` instead of `free_space / 2.0`.

nicoburns commented:
Regarding the most recent WPT test results: I think I accidentally updated the layout2020 expectations rather than the layout2013. I've fixed this in the latest push, but do you know why the parsing tests (more of which are now passing in layout2013) aren't working at all for layout2020?

It seems that querying the `align-content` and `justify-content` properties (and I would guess probably other properties too) with JS isn't returning anything under layout2020 (even though they are affecting the computed layout), which I find surprising as I wouldn't expect this to involve the layout system at all.

Loirooriol commented:
> do you know why the parsing tests (more of which are now passing in layout2013) aren't working at all for layout2020?

That's because `justify-content` and `align-content` have `servo_pref="layout.flexbox.enabled"`.
And this pref is disabled by default, except for tests in `css/css-flexbox/`. The parsing tests are in `css/css-align/`.
Legacy layout not obeying `servo_pref` seems like a bug.

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -328,29 +328,38 @@ impl FlexContainer {
                         }
                     },
-                    AlignContent::Center => free_space / 2.0,
+                    AlignContent::Center => {
+                        if free_space >= Length::zero() {
+                            free_space / 2.0
+                        } else {
+                            Length::zero()
+                            // TODO: implement unsafe fallback: (free_space / 2.0)
+                        }
+                    },
```
`center` should default to unsafe, it was previously fine.

https://drafts.csswg.org/css-align/#typedef-overflow-position

> UAs that have not implemented the “smart” default behavior must behave as [safe](https://drafts.csswg.org/css-align/#valdef-overflow-position-safe) for [align-content](https://drafts.csswg.org/css-align/#propdef-align-content) on [block containers](https://drafts.csswg.org/css-display-4/#block-container) and [unsafe](https://drafts.csswg.org/css-align/#valdef-overflow-position-unsafe) otherwise.

nicoburns replied:
Hmm... the spec for fallbacks (https://drafts.csswg.org/css-align/#distribution-values) seems a little inconsistent:

- `space-between` and `stretch` fallback to `flex-start`
- `space-around` and `space-evenly` fallback to `safe center` which falls back `start` (not `flex-start`)

So `space-between`/`stretch` are going to align to the opposite end of the container to `space-around`/`space-evenly` in the "single alignment subject overflows container" case.

I feel like either `space-between` and `stretch` ought to fallback to `start` or `space-around` and `space-evenly` ought to fallback to `flex-start` and the behaviour of "distributed alignment" keywords in the "single alignment subject" case ought to be defined via a mechanism other than fallback.

Loirooriol replied:
Good point. Seems that Blink makes them fall back to flex-start.

nicoburns commented:
> That's because justify-content and align-content have servo_pref="layout.flexbox.enabled".
And this pref is disabled by default, except for tests in css/css-flexbox/. The parsing tests are in css/css-align/.
Legacy layout not obeying servo_pref seems like a bug.

Ah, that makes sense. I'm thinking it would make sense to enable that pref for `css-align` tests too (I did a quick test this makes a large number of tests pass under layout2020)? I can update this in a follow-up PR too. And perhaps update wpt.servo.org to track `css-align` tests?

Loirooriol commented:
> Legacy layout not obeying servo_pref seems like a bug.

Not a bug, the flag is enabled by default for legacy: https://github.com/servo/servo/blob/1bc63801e79273bd976530af498a43f0b9c7c93b/components/config/opts.rs#L736-L739

nicoburns commented:
@Loirooriol I've updated this such that I believe it now matches the latest spec + the resolution of https://github.com/w3c/csswg-drafts/issues/10154. I have also refactored it so that fallback/safe behaviour is implemented more literally by modifying the alignment mode prior to offsets being computed (so the offsets formulas are all for the "unsafe" variants).

I believe this is ready for another WPT run.

P.S. It seems that with the latest spec `space-between`/``space-around`/`space-evenly`/`stretch` are safe by default but `end`/`center`/etc are not. Is that your interpretation? It wasn't super clear to me from reading the spec, but it seems to be what the WPT tests enforce. It also seems a little inconsistent too, but perhaps it wasn't web compatible to make those safe by default?

nicoburns commented:
@mrobinson I think this also needs a "layout 2013" WPT run

P.S. I've rebased this against latest `main` and latest stylo `main` after the new stylo update was merged yesterday, although as far as I can tell this hasn't caused and changes in the WPT test results.

mrobinson commented:
All of the code changes here are in the layout_2020 directory so it's really unlikely that this has affected results for legacy layout. That said, before landing this, the merge queue will run the legacy layout (layout 2013) WPT suite as well so even if it did change results, it would be caught before landing.

nicoburns commented:
> All of the code changes here are in the layout_2020 directory so it's really unlikely that this has affected results for legacy layout. That said, before landing this, the merge queue will run the legacy layout (layout 2013) WPT suite as well so even if it did change results, it would be caught before landing.

There is also a Stylo change which seems to have changed layout2013 parsing results.

I don't really care if the tests are run or not. My main concern is trying to reduce latency with regard to test runs, such that *if* layout2013 needs to be run I'd rather it be run sooner than later.

nicoburns commented:
@Loirooriol Just a reminder that this is awaiting your re-review in case you have forgotten about it.

Loirooriol reviewed (approved):
Sorry for the delay

mrobinson reviewed (commented):
I've landed the change to stylo so this just needs to update `Cargo.toml` and `Cargo.lock` to land.

nicoburns commented on tests/wpt/meta/css/css-flexbox/align-content-wrap-001.html.ini:
```diff
@@ -6,16 +6,4 @@
-
-  [.default, .verticalWriting 2]
-    expected: FAIL
+    expected: FAIL
```
Yes, the automated update mechanism isn't working very well for me due to platform-specific differences (at least I'm assuming that's the cause). Specifically, the following tests:

CI PASS [local FAIL] /css/css-flexbox/flexbox-flex-direction-column-percentage-ignored.html
CI FAIL [local PASS] /css/css-flexbox/flexbox-mbp-horiz-003-reverse.xhtml
CI FAIL [local PASS] /css/css-flexbox/flexbox-mbp-horiz-003.xhtml
CI FAIL [local PASS] /css/css-flexbox/flexbox-mbp-horiz-003v.xhtml

nicoburns replied:
Newline restored

nicoburns committed:
Upgrade to stylo 6faedad

nicoburns commented:
`Cargo.toml`/`Cargo.lock` updated. I've also applied Oriol's review suggestions.

I've changed the reference in `Cargo.toml` from a `branch` to a `rev` reference (similar to our other git dependencies). This makes it much easier to target a non-latest version of stylo, which is required in this case as servo won't compile with latest stylo until https://github.com/servo/servo/pull/32146 lands). It also makes rebases much easier to process as you can update the commit id in the human-readable Cargo.toml rather than needing to update the Cargo.lock.

mrobinson commented:
> I've changed the reference in `Cargo.toml` from a `branch` to a `rev` reference (similar to our other git dependencies). This makes it much easier to target a non-latest version of stylo, which is required in this case as servo won't compile with latest stylo until #32146 lands). It also makes rebases much easier to process as you can update the commit id in the human-readable Cargo.toml rather than needing to update the Cargo.lock.

This is unnecessary as it can be managed via Cargo.lock and it makes it harder to update stylo as you always have to know what revision to use versus only when you are trying to pin to a specific one. We chose to use the branch name after agreeing among ourselves what the best approach should be, so I think we make changes here based on consensus.

nicoburns commented:
@mrobinson

> This is unnecessary as it can be managed via Cargo.lock and it makes it harder to update stylo as you always have to know what revision to use versus only when you are trying to pin to a specific one. We chose to use the branch name after agreeing among ourselves what the best approach should be, so I think we make changes here based on consensus.

Hmm... ok sorry. Well whoever next bumps stylo can switch it back.

Happy to go with the consensus, but I'm not personally convinced that the chosen approach is easier: specifically in this case (and it seems to me that it could easily happen in future) the required commit is not represented by a branch (because the `main` and `2024-04-16` that it has been merged to both have a more recent commit).

Is there an easy way to update Cargo.lock to a specific commit of a branch without temporarily specifying it in the Cargo.toml and then changing it back or manually updating the commit ids in the Cargo.lock (which requires you to know both the old and new commit id).

mrobinson commented:
> Is there an easy way to update Cargo.lock to a specific commit of a branch without temporarily specifying it in the Cargo.toml and then changing it back or manually updating the commit ids in the Cargo.lock (which requires you to know both the old and new commit id).

My workflow generally to switch Servo's `Cargo.toml` to the branch with the Stylo changes and then when they land in Stylo, I undo those `Cargo.toml` changes and run `./mach build` again. I haven't had any issues with this approach so far and I think other folks are using it to.

nicoburns commented:
> My workflow generally to switch Servo's `Cargo.toml` to the branch with the Stylo changes and then when they land in Stylo, I undo those `Cargo.toml` changes and run `./mach build` again. I haven't had any issues with this approach so far and I think other folks are using it to.

IIRC that is what I tried first, but I ended up with compile errors because `branch = "main"` means the latest commit of the `main` branch and this PR / servo `main` won't currently compile with latest `stylo` main. And in general this will be case any time more than one breaking change is landed to stylo at once (before the corresponding servo PRs have been landed).

mrobinson commented:
> IIRC that is what I tried first, but I ended up with compile errors because `branch = "main"` means the latest commit of the `main` branch and this PR / servo `main` won't currently compile with latest `stylo` main. And in general this will be case any time more than one breaking change is landed to stylo at once (before the corresponding servo PRs have been landed).

Instead of switching branch to be `main` switch it to be your WIP branch in Stylo or modify the line in `Cargo.toml` to reference something like `rev = "refs/pull/493/head"` as described [in the documentation](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).

In the case that multiple changes land in Stylo (that need updates in Servo) and without Servo being updated in between (this has never happened for me, but can happen), you will then need to set `Cargo.lock` to track a particular revision in order to land the changes in Servo. I think this just means passing a precise revision to `cargo update` before running `./mach build`.

Loirooriol commented:
I don't particularly mind changing the approach, but probably this conversation should happen in a https://github.com/servo/stylo/ issue.

For now in #32146 I'm going back to using the branch.
